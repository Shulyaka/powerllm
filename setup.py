"""Setup module for powerllm."""

import pathlib

from setuptools import find_packages, setup

VERSION = "0.0.1"


def long_description():
    """Read README.md file."""
    f = (pathlib.Path(__file__).parent / "README.md").open()
    res = f.read()
    f.close()
    return res  # noqa: R504


setup(
    name="powerllm",
    version=VERSION,
    description="Home Assistant custom component " "to empower LLM integrations",
    long_description=long_description(),
    long_description_content_type="text/markdown",
    url="https://github.com/Shulyaka/powerllm",
    author="Denis Shulyaka",
    author_email="ds_github@shulyaka.org.ru",
    license="GNU General Public License v3.0",
    keywords="homeassistant home-assistant llm generative ai chatgpt chat gpt gemini",
    packages=find_packages(exclude=["tests"]),
    python_requires=">=3.12",
    tests_require=["pytest"],
)
