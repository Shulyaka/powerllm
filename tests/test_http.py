"""Tests for LLM Tools HTTP API."""

import voluptuous as vol
from homeassistant.core import HomeAssistant
from homeassistant.helpers import config_validation as cv, intent
from homeassistant.setup import async_setup_component
from pytest_homeassistant_custom_component.common import MockUser
from pytest_homeassistant_custom_component.typing import ClientSessionGenerator
from syrupy.assertion import SnapshotAssertion


def test_test(hass):
    """Workaround for https://github.com/MatthewFlamm/pytest-homeassistant-custom-component/discussions/160."""


async def test_http_api_list(
    hass: HomeAssistant,
    hass_client: ClientSessionGenerator,
    hass_admin_user: MockUser,
) -> None:
    """Test LLM API list via HTTP API."""
    assert await async_setup_component(hass, "powerllm", {})
    client = await hass_client()
    resp = await client.get("/api/powerllm")

    assert resp.status == 200
    data = await resp.json()

    assert data == [{"name": "Assist", "id": "assist"}]


async def test_http_tool_list(
    hass: HomeAssistant,
    hass_client: ClientSessionGenerator,
    hass_admin_user: MockUser,
    snapshot: SnapshotAssertion,
) -> None:
    """Test LLM Tool list via HTTP API."""
    assert await async_setup_component(hass, "powerllm", {})

    class TestIntentHandler(intent.IntentHandler):
        """Test Intent Handler."""

        intent_type = "OrderBeer"
        description = "Orders beer"

        @property
        def slot_schema(self) -> dict | None:
            """Return a slot schema."""
            return {vol.Required("type"): cv.string}

        async def async_handle(self, intent):
            """Handle the intent."""
            assert intent.context.user_id == hass_admin_user.id
            response = intent.create_response()
            response.async_set_speech(
                f"I've ordered a {intent.slots['type']['value']}!"
            )
            response.async_set_card(
                "Beer ordered", f"You chose a {intent.slots['type']['value']}."
            )
            return response

    intent.async_register(hass, TestIntentHandler())

    client = await hass_client()
    resp = await client.get("/api/powerllm/assist")

    assert resp.status == 200
    data = await resp.json()

    data["tools"][0]["parameters"]["properties"]["device_class"]["items"]["enum"].sort()
    data["tools"][1]["parameters"]["properties"]["device_class"]["items"]["enum"].sort()
    assert data == snapshot

    resp = await client.get("/api/powerllm/non-existent")
    assert resp.status == 404

    resp = await client.post(
        "/api/powerllm/assist",
        json={"tool_name": "OrderBeer", "tool_args": {"type": "Belgian"}},
    )

    assert resp.status == 200
    data = await resp.json()

    assert data == {
        "data": {
            "failed": [],
            "success": [],
            "targets": [],
        },
        "response_type": "action_done",
        "speech": {
            "plain": {
                "extra_data": None,
                "speech": "I've ordered a Belgian!",
            },
        },
    }

    resp = await client.post(
        "/api/powerllm/non-existent", json={"tool_name": "OrderBeer"}
    )
    assert resp.status == 404

    resp = await client.post("/api/powerllm/assist", json={"tool_name": "OrderBeer"})
    assert resp.status == 500
    data = await resp.json()

    # assert data == {
    #    "error": "MultipleInvalid",
    #    "error_text": "required key not provided @ data['type']",
    # }
    assert data == {
        "error": "IntentUnexpectedError",
        "error_text": "Error handling OrderBeer",
    }


async def test_http_tool(
    hass: HomeAssistant, hass_client: ClientSessionGenerator, hass_admin_user: MockUser
) -> None:
    """Test LLM Tool via HTTP API."""
    assert await async_setup_component(hass, "powerllm", {})

    class TestIntentHandler(intent.IntentHandler):
        """Test Intent Handler."""

        intent_type = "OrderBeer"
        description = "Orders beer"

        @property
        def slot_schema(self) -> dict | None:
            """Return a slot schema."""
            return {vol.Required("type"): cv.string}

        async def async_handle(self, intent):
            """Handle the intent."""
            assert intent.context.user_id == hass_admin_user.id
            response = intent.create_response()
            response.async_set_speech(
                f"I've ordered a {intent.slots['type']['value']}!"
            )
            response.async_set_card(
                "Beer ordered", f"You chose a {intent.slots['type']['value']}."
            )
            return response

    intent.async_register(hass, TestIntentHandler())

    client = await hass_client()
    resp = await client.get("/api/powerllm/assist/OrderBeer")

    assert resp.status == 200
    data = await resp.json()

    assert data == {
        "description": "Orders beer",
        "name": "OrderBeer",
        "parameters": {
            "properties": {
                "type": {
                    "type": "string",
                },
            },
            "required": ["type"],
            "type": "object",
        },
    }

    resp = await client.get("/api/powerllm/non-existent/non-existent")
    assert resp.status == 404

    resp = await client.get("/api/powerllm/assist/non-existent")
    assert resp.status == 404

    resp = await client.post("/api/powerllm/assist/OrderBeer", json={"type": "Lager"})

    assert resp.status == 200
    data = await resp.json()

    assert data == {
        "data": {
            "failed": [],
            "success": [],
            "targets": [],
        },
        "response_type": "action_done",
        "speech": {
            "plain": {
                "extra_data": None,
                "speech": "I've ordered a Lager!",
            },
        },
    }

    resp = await client.post("/api/powerllm/non-existent/non-existent")
    assert resp.status == 404

    resp = await client.post("/api/powerllm/assist/non-existent")
    assert resp.status == 404

    resp = await client.post("/api/powerllm/assist/OrderBeer")

    assert resp.status == 500
    data = await resp.json()

    # assert data == {
    #    "error": "MultipleInvalid",
    #    "error_text": "required key not provided @ data['type']",
    # }
    assert data == {
        "error": "IntentUnexpectedError",
        "error_text": "Error handling OrderBeer",
    }
