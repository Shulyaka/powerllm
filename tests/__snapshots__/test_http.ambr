# serializer version: 1
# name: test_http_tool_list
  dict({
    'prompt': 'Only if the user wants to control a device, tell them to expose entities to their voice assistant in Home Assistant.',
    'tools': list([
      dict({
        'description': 'Turns on/opens a device or entity',
        'name': 'HassTurnOn',
        'parameters': dict({
          'properties': dict({
            'area': dict({
              'type': 'string',
            }),
            'device_class': dict({
              'items': dict({
                'enum': list([
                  'awning',
                  'blind',
                  'curtain',
                  'damper',
                  'door',
                  'garage',
                  'gas',
                  'gate',
                  'outlet',
                  'receiver',
                  'shade',
                  'shutter',
                  'speaker',
                  'switch',
                  'tv',
                  'water',
                  'window',
                ]),
                'type': 'string',
              }),
              'type': 'array',
            }),
            'domain': dict({
              'items': dict({
                'type': 'string',
              }),
              'type': 'array',
            }),
            'floor': dict({
              'type': 'string',
            }),
            'name': dict({
              'type': 'string',
            }),
          }),
          'required': list([
          ]),
          'type': 'object',
        }),
      }),
      dict({
        'description': 'Turns off/closes a device or entity',
        'name': 'HassTurnOff',
        'parameters': dict({
          'properties': dict({
            'area': dict({
              'type': 'string',
            }),
            'device_class': dict({
              'items': dict({
                'enum': list([
                  'awning',
                  'blind',
                  'curtain',
                  'damper',
                  'door',
                  'garage',
                  'gas',
                  'gate',
                  'outlet',
                  'receiver',
                  'shade',
                  'shutter',
                  'speaker',
                  'switch',
                  'tv',
                  'water',
                  'window',
                ]),
                'type': 'string',
              }),
              'type': 'array',
            }),
            'domain': dict({
              'items': dict({
                'type': 'string',
              }),
              'type': 'array',
            }),
            'floor': dict({
              'type': 'string',
            }),
            'name': dict({
              'type': 'string',
            }),
          }),
          'required': list([
          ]),
          'type': 'object',
        }),
      }),
      dict({
        'description': 'Cancels all timers',
        'name': 'HassCancelAllTimers',
        'parameters': dict({
          'properties': dict({
            'area': dict({
              'type': 'string',
            }),
          }),
          'required': list([
          ]),
          'type': 'object',
        }),
      }),
      dict({
        'description': 'Orders beer',
        'name': 'OrderBeer',
        'parameters': dict({
          'properties': dict({
            'type': dict({
              'type': 'string',
            }),
          }),
          'required': list([
            'type',
          ]),
          'type': 'object',
        }),
      }),
    ]),
  })
# ---
